service/matching/version_rule_helpers.go

```diff
@@ -26,219 +26,187 @@ package matching
 	if actualIdx < 0 {
-		return nil, serviceerror.NewInvalidArgument(fmt.Sprintf("rule index %d is out of bounds for assignment rule list of length %d", idx, len(getActiveRules(rules))))
+		return nil, serviceerror.NewInvalidArgument(fmt.Sprintf(
+			"rule index %d is out of bounds for assignment rule list of length %d", idx, len(getActiveAssignmentRules(rules))))
 	}
-	persistenceAR := persistencepb.AssignmentRule{
+	rules[actualIdx].DeleteTimestamp = timestamp
+	data.AssignmentRules = slices.Insert(rules, actualIdx, &persistencepb.AssignmentRule{
 		Rule:            rule,
 		CreateTimestamp: timestamp,
 		DeleteTimestamp: nil,
-	}
-	slices.Replace(data.AssignmentRules, actualIdx, actualIdx+1, &persistenceAR)
-	return data, checkAssignmentConditions(data, 0, req.GetForce(), hadUnfiltered)
+	})
+	return data, checkAssignmentConditions(data, 0, hadUnfiltered && !req.GetForce())
 }
 
 func DeleteAssignmentRule(timestamp *hlc.Clock,
 	data *persistencepb.VersioningData,
 	req *workflowservice.UpdateWorkerVersioningRulesRequest_DeleteBuildIdAssignmentRule,
-	hadUnfiltered bool) (*persistencepb.VersioningData, error) {
+) (*persistencepb.VersioningData, error) {
 	data = common.CloneProto(data)
 	rules := data.GetAssignmentRules()
+	hadUnfiltered := containsUnfiltered(rules)
 	idx := req.GetRuleIndex()
 	actualIdx := given2ActualIdx(idx, rules)
-	if actualIdx < 0 {
-		return nil, serviceerror.NewInvalidArgument(fmt.Sprintf("rule index %d is out of bounds for assignment rule list of length %d", idx, len(getActiveRules(rules))))
+	if actualIdx < 0 || actualIdx > len(rules)-1 {
+		return nil, serviceerror.NewInvalidArgument(fmt.Sprintf(
+			"rule index %d is out of bounds for assignment rule list of length %d", idx, len(getActiveAssignmentRules(rules))))
 	}
-	rule := rules[actualIdx]
-	rule.DeleteTimestamp = timestamp
-	return data, checkAssignmentConditions(data, 0, req.GetForce(), hadUnfiltered)
-}
-
-// checkRedirectConditions returns an error if there is no room for the new Redirect Rule, or if it causes the
-// queue to fail the other requirements
-func checkRedirectConditions(g *persistencepb.VersioningData, maxRRs int) error {
-	return nil
-}
-
-func countActiveRR(rules []*persistencepb.RedirectRule) int {
-	return 0
-}
-
-// trimRedirectRules attempts to trim the DAG of redirect rules. It returns the number of rules it was able to delete.
-func trimRedirectRules(rules []*persistencepb.RedirectRule) int {
-	return 0
-}
-
-func isCyclic(rules []*persistencepb.RedirectRule) bool {
-	return false
+	rules[actualIdx].DeleteTimestamp = timestamp
+	return data, checkAssignmentConditions(data, 0, hadUnfiltered && !req.GetForce())
 }
 
 func InsertCompatibleRedirectRule(timestamp *hlc.Clock,
 	data *persistencepb.VersioningData,
 	req *workflowservice.UpdateWorkerVersioningRulesRequest_AddCompatibleBuildIdRedirectRule,
-	maxRRs int) (*persistencepb.VersioningData, error) {
-	return nil, nil
+	maxRedirectRules int) (*persistencepb.VersioningData, error) {
```

Comment:
do we check source and target not being the same? or it's gonna be caught by isCyclic?

Reply:
and `TestReplaceRedirectRuleCreateCycle`